### ISSUE TITLE
Missing GZIP compression support of API endpoints

### ISSUE THREAD

**Issue by jo-me** [valhalla/valhalla#4298]
Hi,

it seems like many if not all API endpoints do not support GZIP compression out-of-the-box, e.g. the route endpoint or trace_attributes.

Is this assumption correct?

If so, is there a way to enable this or what is the suggested procedure to support this for clients?

Thanks!

Jochen Mehlhorn <jochen.mehlhorn@mercedes-benz.com>, Mercedes-Benz Tech Innovation GmbH

[Provider Information](https://github.com/Daimler/daimler-foss/blob/master/PROVIDER_INFORMATION.md)

**Comment by ianthetechie** [comment]
I know this isn't the exact question you asked @jo-me, but out of curiosity why not let your reverse proxy handle this? *Most* (though you may have different constraints) API deployment architectures will want some sort of reverse proxy like HAProxy or nginx in front, and those have a well tested gzip implementation already.

**Comment by jo-me** [comment]
You're right @ianthetechie, this is a possible option..  if there is no better (native) way. 

I would consider it better to have gzip support as close to the source of the data as possible because adding compression in a reverse proxy may result in add additonal network hop(s) for the uncompressed data. 

Can I interpret the thumbsup to your response as a confirmation that this is the preferred way to go about this?

**Comment by nilsnolde** [comment]
> Can I interpret the thumbsup to your response as a confirmation that this is the preferred way to go about this?

Fair question 😄 It’d surely be nicer to have our own gzip compression. Maybe it’d be a better fit for https://github.com/kevinkreiser/prime_server as that’s the server part of Valhalla. 

If you’re worried about response size, you can already request pbf format for most endpoints. It’s also faster as a side effect since that’s our internal format anyways so you skip the whole JSON serialization.

**Comment by ianthetechie** [comment]
Yeah, PBF would remove that size and serialization question. Though that assumes your clients would be able to consume that (may not be an option if you want OSRM format, for example).

Let me clarify my suggestion a bit more, since I realize it was ambiguous now ;) There are multiple ways you can use reverse proxies. If you are concerned about the speed of the link transmitting data uncompressed over your "internal" network before your "external" facing reverse proxy, you can run HAProxy on the same machine where the Valhalla instance lives. You can still get a great gzip implementation, you protect Valhalla from (accidental or intentional) malformed HTTP requests by not exposing its port publicly. HAProxy is extremely light, so this should be perfectly fine in every situation I can think of.

I also agree moving compression to prime_server if desired probably makes most sense vs adding it to individual Valhalla endpoints.

**Comment by kevinkreiser** [comment]
yes this is a feature request for prime_server and yes you can get the same functionality out of whatever you have in front of that, eg reverse proxy.

**Comment by kevinkreiser** [comment]
https://github.com/kevinkreiser/prime_server/issues/129

**Comment by jo-me** [comment]
Alright understood, thank you all for the clarifications
